README.md
# NAME

epan - Exclusive Perl Archive Nook

# VERSION

Ask the version number to the script itself, calling:

    shell$ epan --version

# USAGE

    epan [--usage] [--help] [--man] [--version]

    # "create" demands that dirname is not present, to create it
    epan create
       [-1|-m|--mailrc path]
       [-2|-o|--output path]
       [-3|-l|--modlist path]
       [-t|--target dirname]
       Module1 [Module2...]

    # "index" is also "idx"
    epan index
       [-1|-m|--mailrc path]
       [-2|-o|--output path]
       [-3|-l|--modlist path]
       [-t|--target dirname]

    # "inject" adds local distribution archives available as files
    epan inject
       [-1|-m|--mailrc path]
       [-2|-o|--output path]
       [-3|-l|--modlist path]
       [-a|--author name]
       [-t|--target dirname]
       File1 [File2...]

    # "list-actions" is also "list_actions"
    epan list-actions

    # "list-obsoletes" is also "list_obsoletes"
    epan list-obsoletes [-t|--target dirname]

    # "purge-obsoletes" is also "purge_obsoletes"
    epan purge-obsoletes [-t|--target dirname]

    # "update" is also "add" and "install"
    epan update
       [-1|-m|--mailrc path]
       [-2|-o|--output path]
       [-3|-l|--modlist path]
       [-t|--target dirname]
       Module1 [Module2...]

# EXAMPLES

    # collects all what's needed to install Template::Perlish somewhere
    shell$ epan create -t mymodules Template::Perlish

    # regenerate index in ./modules/02packages.details.txt.gz
    shell$ epan idx -t mymodules

    # prints index on standard output, works on /path/to/minicpan
    shell$ epan index -o - -t /path/to/minicpan

# DESCRIPTION

This program helps you creating and managing an EPAN - a version of the
CPAN that is trimmed down to your needs for installing specific stuff.

To start with an example, suppose you have to install Mojolicious and
a couple of its plugins in a machine that - for good reasons - is not
connected to the Internet. It's easy to get the distribution files for
Dancer and the plugins... but what about the dependencies? It can easily
become a nightmare, forcing you to go back and forth with new modules as
soon as you discover the need to install them.

Thanks to [cpanm](https://metacpan.org/pod/cpanm), this is quite easier these days: it can actually do
what's needed with a single command:

    # on the machine connected to the Internet or to a minicpan
    $ cpanm -L xxx --scandeps --save-dists dists \
         Mojolicious Mojolicious::Plugin::Authentication IO::Socket::SSL ...

which places all the modules in subdirectory `dists` (thanks to option
`--save-dists`) with an arrangement similar to what you would expect from
a CPAN mirror.

On the target machine, you still have to make some work - e.g. you should
collect the output from the invocation of cpanm above to figure out the
order to use for installing the distribution files. Additionally, the
directory structure that is generated lacks a proper index file (located
in `modules/02package.details.txt.gz`) so it would be difficult to use
the normal toolchain.

[epan](https://metacpan.org/pod/epan) aims at filling up the last mile to get the job done, providing
you with a subdirectory that is ready for deployment, with all the bits in
place to push automation as much as possible. So you can do this:

    # on the machine connected to the Internet or to a minicpan
    $ epan create Mojolicious Mojolicious::Plugin::Authentication \
        IO::Socket::SSL ...
    $ tar cvzf epan.tar.gz epan

transfer `dists.tar.gz` to the target machine and...

    # on the target machine
    $ tar xvzf epan.tar.gz
    $ cd epan
    $ ./install.sh

optionally providing an installation target directory:

    $ ./install.sh /path/to/local/perl

The program `epan` is actually a unified access point to several
different tools for manipulating your _exclusive Perl archive nook_. Most
of these commands operate upon a _target directory_ that is where your
EPAN is stored; this can be specified via option `-t` or its longer
version `--target`. By default, the target directory is assumed to be
`epan` in the current directory.

## `add`, `install` and `update`

These commands are synonimous in `epan`, and all help you pull a module
and its dependencies from a CPAN mirror right into your EPAN, regenerating
the index at the end of the process. The syntax is:

    epan add # or install or update \
       [-t|--target directory]
       Module1 [Module2...]

So, in addition to the common option `-t` for setting the right target
directory, it accepts a list of module names to install (with their
dependencies).

## `create` 

This command is almost the same as `add` and its aliases, with the
exception that the target directory MUST NOT already exist when called.

## `index`

Regenerate the index so that tools like `cpanm` are happy about what they
find and treat your target directory as a real CPAN sort-of mirror. The
syntax is the following:

    epan index [-t|--target dirname]

Note that other commands (e.g. `add` or `create`) already do the
indexing. This command can be useful when you have a starting base (i.e.
a compound of modules coming from CPAN and your own distribution) already
arranged in the right directory tree, but you need to generate an index.
For example, this happens when you collect some distribution files using
`cpanm`:

    shell$ cpanminus -L xxx --save-dists dists Mod1 Mod2...

because it saves the needed distributions in `dists` but it does not
generate the index. The same happens when using `carton`.

In these cases, if you want to prepare a pack of modules to carry with your
application, you can do like this:

    $ figure_out_modules > modlist
    $ cpanm -L xxx --save-dists dists $(<modlist)
    $ epan index -t dists
    $ tar cvf dists.tar dists

Well, put like this the second and third lines can just be synthesized as:

    $ epan add -t dists $(<modlist)

but you get the idea. The directory with the modules might be the
byproduct of invoking `carton` instead (in which case you would end up
with a sub-directory `cache`).

Anyway, you can then carry dists.tar with you, at which point you can:

    $ cpanm --mirror file://$YOURPATH --mirror-only Mod1 Mod2 ...

This command expects the target directory to exist and will complain
otherwise.

## `inject`

If you have some local distribution files, e.g. generated by yourself and
not (yet) uploaded to CPAN, you can inject them into a local EPAN. The
syntax is straightforward:

    epan inject \ 
       [-a|--author author-name] \
       [-t|--target dirname] File1 [File2...]

As for many other commands, you can also set the different output
filenames, but that would be hardly useful.

## `list-actions` and `list_actions`

Prints out the list of available commands.

## `list-obsoletes` and `list_obsoletes`

    epan list-obsoletes [-t|--target dirname]

Prints out a list of obsolete distributions in the EPAN. A distribution is
considered _obsolete_ if there is a newer corresponding version in the
EPAN. E.g. suppose that you work on `Acme::Whatever` and inject version
`0.2`:

    epan inject Acme-Whatever-0.2.tar.gz
    # ...

then you work on it some more time, and inject version `0.3`:

    epan inject Acme-Whatever-0.2.tar.gz
    # ...

Now your EPAN contains two distribution packages for `Acme::Whatever`,
one for release `0.2` (which is the obsolete one) and one for the newest
version `0.3`.

## `purge-obsoletes` and `purge_obsoletes`

    epan purge-obsoletes [-t|--target dirname]

Remove (purge) obsolete distribution packages from the EPAN. See above for
what _obsolete_ means.

# OPTIONS

The following options are supported, even though not all actions use them
all:

- -1 | -m | --mailrc

    path to the file `01mailrc.txt.gz`, defaults to
    `authors/01mailrc.txt.gz` inside the target directory. You hardly want to
    change this.

- -2 | -o | --output | --package-details

    path to the file for `02packages.details.txt.gz`, defaults to
    `modules/02packages.details.txt.gz` inside the target directory. Yes, you
    can use `-` with the _usual_ meaning, although this might not help you
    much.

- -3 | -l | --modlist | --modlist-data

    path to the file `03modlist.data.gz`, defaults to
    `modules/03modlist.data.gz` inside the target directory. You hardly want
    to change this.

- -a | --author author-name

    module author to use when doing injection of local distribution packages.

- --help

    print a somewhat more verbose help, showing usage, this description of the
    options and some examples.

- --man

    print out the full documentation for the script.

- -t | --target dirname

    set the directory of the root for the EPAN to work on. Defaults to the
    sub-directory `epan` in the current directory. This option applies to all
    commands except `list-actions`.

    The default value is `epan` as a sub-directory in the current directory.

    Some commands demand that the target directory already exists; in case it
    does not, they will complain that you're probably using this option in the
    wrong way (or not using it at all).

- --usage

    print a concise usage line and exit.

- --version

    print the version of the script.

# CONFIGURATION AND ENVIRONMENT

epan requires no configuration files. The following environment variable
is honored:

- `EPAN_AUTHOR`

    set the name of the _pause account_ to use for indexing. Defaults to
    `LOCAL`. It is overridden by `--author`.

# DEPENDENCIES

Runs on perl 5.012, adapt it if you want to run on something older :-)

The following non-core modules are used:

- **Dist::Metadata**
- **Path::Class**
- **File::Find::Rule**
- **Log::Log4perl::Tiny**

# BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests through
[https://github.com/polettix/epan/issues](https://github.com/polettix/epan/issues).

# AUTHOR

Flavio Poletti `polettix@cpan.org`

# LICENCE AND COPYRIGHT

Copyright (C) 2011-2021 by Flavio Poletti `polettix@cpan.org`.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

> [http://www.apache.org/licenses/LICENSE-2.0](http://www.apache.org/licenses/LICENSE-2.0)

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


Changes
{{$NEXT}}

0.002  2021-06-19 11:26:27 CEST
   - some refactoring
   - documentation fixes

0.001002  2021-06-14 15:47:32 CEST
   - use hint about ignoring CPAN::Modulelist via CPAN::Meta stuff

0.001001  2021-06-12 18:23:27 CEST
   - avoid tripping PAUSE on a text with a package inside   

0.001  2021-06-12 17:59:55 CEST
   - initial packaging


lib/App/EPAN.pm
package App::EPAN;

use 5.012;
{ our $VERSION = '0.001002' }
use warnings;
use English qw( -no_match_vars );
use version;
use autodie;
use Getopt::Long qw< :config gnu_getopt >;
use Pod::Usage qw< pod2usage >;
use Dist::Metadata ();
use Path::Class qw< file dir >;
use Cwd qw< cwd >;
use File::Find::Rule ();
use Compress::Zlib   ();
use Log::Log4perl::Tiny qw< :easy :dead_if_first >;
use Moo;
use IPC::Run   ();
use File::Copy ();
use File::Which qw< which >;

has configuration => (
   is        => 'rw',
   lazy      => 1,
   predicate => 'has_config',
   clearer   => 'clear_config',
   default   => sub { {} },
);
has action     => (is => 'rw',);
has last_index => (is => 'rw',);

sub run {
   my $package = shift;
   my $self    = $package->new();
   $self->get_options(@_);

   my $action = $self->action();
   pod2usage(-verbose => 99, -sections => 'USAGE') unless defined $action;
   if (my $method = $self->can("action_$action")) {
      $self->$method();
   }
   else {
      FATAL "action '$action' is not supported\n";
      $self->action_list_actions;
      exit 1;
   }
   return;
} ## end sub run

sub get_options {
   my $self = shift;
   my $action =
     (scalar(@_) && length($_[0]) && (substr($_[0], 0, 1) ne '-'))
     ? shift(@_)
     : 'list-actions';
   $action =~ s{-}{_}gmxs;
   local @ARGV = @_;
   $self->action($action);
   my %config = ();
   GetOptions(
      \%config,
      qw(
        mailrc|m|1=s
        output|packages-details|o|2=s
        modlist|modlist-data|l|3=s
        target|t=s
        test|T!
        author|a=s
        usage! help! man! version!
        )
   ) or pod2usage(-verbose => 99, -sections => 'USAGE');
   our $VERSION ||= 'whateva';
   pod2usage(message => "$0 $VERSION", -verbose => 99, -sections => ' ')
     if $config{version};
   pod2usage(-verbose => 99, -sections => 'USAGE') if $config{usage};
   pod2usage(-verbose => 99, -sections => 'USAGE|EXAMPLES|OPTIONS')
     if $config{help};
   pod2usage(-verbose => 2) if $config{man};
   $self->configuration(
      {
         cmdline_config => \%config,
         config         => \%config,
         args           => [@ARGV],
      }
   );
   return;
} ## end sub get_options

sub args {
   return @{$_[0]->configuration()->{args}};
}

sub config {
   my $self = shift;
   return @{$self->configuration()->{config}}{@_} if wantarray();
   return $self->configuration()->{config}{shift @_};
}

sub target_dir {
   my $self = shift;
   return dir($self->config('target') // 'epan');
}

sub execute_tests {
   my $self = shift;
   return $self->config('test');
}

sub action_index { return shift->_do_index }

{
   no strict 'refs';
   *{action_idx} = \&action_index;
}

sub _save {
   my ($self, $name, $contents, $config_key, $output) = @_;

   if (defined(my $confout = $self->config($config_key))) {
      $output =
          !length($confout) ? undef
        : $confout eq '-'   ? \*STDOUT
        :                     file($confout);
   } ## end if (defined(my $confout...))
   if (defined $output) {
      INFO "saving output to $output";
      $self->_save2($output,
         scalar(ref($contents) ? $contents->() : $contents));
   }
   else {
      INFO "empty filename for $name file, skipping";
   }
} ## end sub _save

sub _do_index {
   my ($self, $basedir) = @_;
   $basedir //= $self->target_dir;
   LOGDIE "path '$basedir' does not exist (wrong -t option?)"
      unless -d $basedir;

   $self->_save(
      '01mailrc',    # name
      '',            # contents
      'mailrc',      # configuration key to look output file
      $basedir->file(qw< authors 01mailrc.txt.gz >)    # default
   );

   $self->_save(
      '02packages.details',    # name
      sub {                    # where to get data from. Call is avoided if
                               # no file on output
         INFO "getting contributions for regenerated index...";
         $self->_index_for($basedir);
      },
      'output',                # configuration key to look output file
      $basedir->file(qw< modules 02packages.details.txt.gz >)    # default
   );

   $self->_save(
      '03modlist.data',                                          # name
      <<'END_OF_03_MODLIST_DATA',
File:        03modlist.data
Description: These are the data that are published in the module
        list, but they may be more recent than the latest posted
        modulelist. Over time we'll make sure that these data
        can be used to print the whole part two of the
        modulelist. Currently this is not the case.
Modcount:    0
Written-By:  PAUSE version 1.005
Date:        Sun, 28 Jul 2013 07:41:15 GMT

package CPAN::Modulelist;
# Usage: print Data::Dumper->new([CPAN::Modulelist->data])->Dump or similar
# cannot 'use strict', because we normally run under Safe
# use strict;
sub data {
   my $result = {};
   my $primary = "modid";
   for (@$CPAN::Modulelist::data){
      my %hash;
      @hash{@$CPAN::Modulelist::cols} = @$_;
      $result->{$hash{$primary}} = \%hash;
   }
   return $result;
}
$CPAN::Modulelist::cols = [ ];
$CPAN::Modulelist::data = [ ];
END_OF_03_MODLIST_DATA
      'modlist',    # configuration key to look output file
      $basedir->file(qw< modules 03modlist.data.gz >)    # default
   );
} ## end sub _do_index

sub _save2 {
   my ($self, $path, $contents) = @_;
   my ($fh, $is_gz);
   if (ref($path) eq 'GLOB') {
      $fh    = $path;
      $is_gz = 0;
   }
   else {
      $path->dir()->mkpath() unless -d $path->dir()->stringify();
      $fh    = $path->open('>');
      $is_gz = $path->stringify() =~ m{\.gz$}mxs;
   }

   if ($is_gz) {
      my $gz = Compress::Zlib::gzopen($fh, 'wb');
      $gz->gzwrite($contents);
      $gz->gzclose();
   }
   else {
      binmode $fh;
      print {$fh} $contents;
   }
   return;
} ## end sub _save2

sub _index_for {
   my ($self, $path) = @_;
   $path //= $self->target_dir;
   my @index = $self->_index_body_for($path);
   our $VERSION ||= 'whateva';
   my $header = <<"END_OF_HEADER";
File:         02packages.details.txt
URL:          http://cpan.perl.org/modules/02packages.details.txt.gz
Description:  Package names found in directory \$CPAN/authors/id/
Columns:      package name, version, path
Intended-For: Automated fetch routines, namespace documentation.
Written-By:   epan $VERSION
Line-Count:   ${ \ scalar @index }
Last-Updated: ${ \ scalar localtime() }
END_OF_HEADER
   return join "\n", $header, @index, '';
} ## end sub _index_for

sub _collect_index_for {
   my ($self, $path) = @_;
   $path //= $self->target_dir;
   $path = dir($path);
   LOGDIE "path '$path' does not exist (wrong -t option?)" unless -d $path;

   my $idpath = $path->subdir(qw< authors id >);
   my %data_for;
   for my $file (File::Find::Rule->extras({follow => 1})->file()
      ->in($idpath->stringify()))
   {
      INFO "indexing $file";
      my $index_path =
        file($file)->relative($idpath)->as_foreign('Unix')->stringify();
      my $dm = Dist::Metadata->new(file => $file);
      my $version_for = $dm->package_versions();

      $data_for{distro}{$index_path} = $version_for;
      (my $bare_index_path = $index_path) =~
        s{^(.)/(\1.)/(\2.*?)/}{$3/}mxs;
      $data_for{bare_distro}{$bare_index_path} = $version_for;

      my %_localdata_for;
      my $score = 0;
      my $previous;
      while (my ($module, $version) = each %$version_for) {
         my $print_version = $version // 'undef';
         DEBUG "data for $module: [$print_version] [$index_path]";
         $_localdata_for{$module} = {
            version => $version,
            distro  => $index_path,
            _file   => $file,
         };
         next if $score != 0;
         next unless exists($data_for{module}{$module});
         $previous = $data_for{module}{$module};
         DEBUG 'some previous version exists';
         if (! defined $version) {
            $score = -1 if defined($previous->{version});
         }
         elsif (defined $previous->{version}) {
            my $tv = version->parse($version);
            my $pv = version->parse($previous->{version});
            $score = $tv <=> $pv;
         }
         DEBUG "score: $score";
      } ## end while (my ($module, $version...))

      DEBUG "FINAL SCORE $score";

      if ($score < 0) { # didn't win against something already in
         DEBUG "marking $file as obsolete";
         $data_for{obsolete}{$file} = 1;
         next;
      }

      DEBUG "getting $file data as winner (for the moment)";
      if ($previous) {
         my $oip = $previous->{distro};
         DEBUG "marking $oip as obsolete";
         $data_for{obsolete}{$previous->{_file}} = 1;
         delete $data_for{module}{$_}
           for keys %{$data_for{distro}{$oip}};
      }
      # copy stuff over to the "official" data for modules
      $data_for{module}{$_} = $_localdata_for{$_} for keys %_localdata_for;
   } ## end for my $file (File::Find::Rule...)
   $self->last_index(\%data_for);
   return %data_for if wantarray();
   return \%data_for;
} ## end sub _collect_index_for

sub _index_body_for {
   my ($self, $path) = @_;
   $path //= $self->target_dir;

   my $data_for        = $self->_collect_index_for($path);
   my $module_data_for = $data_for->{module};
   my @retval;
   for my $module (sort keys %{$module_data_for}) {
      my $md         = $module_data_for->{$module};
      my $version    = $md->{version} || 'undef';
      my $index_path = $md->{distro};
      my $fw         = 38 - length $version;
      $fw = length $module if $fw < length $module;
      push @retval, sprintf "%-${fw}s %s  %s", $module, $version,
        $index_path;
   } ## end for my $module (sort keys...)
   return @retval if wantarray();
   return \@retval;
} ## end sub _index_body_for

sub action_create {
   my ($self) = @_;

   my $target = $self->target_dir;
   LOGDIE "target directory $target exists, use update instead"
     if -d $target;
   $target->mkpath();

   return $self->action_update;
} ## end sub action_create

sub action_update {
   my ($self) = @_;

   my $target = $self->target_dir;
   $target->mkpath() unless -d $target;

   my $dists   = $target->stringify();
   my $local   = $target->subdir('local')->stringify();
   my @command = (
      qw< cpanm --reinstall --quiet --self-contained >,
      ($self->execute_tests ? () : '--notest'),
      '--local-lib-contained' => $local,
      '--save-dists'          => $dists,
      $self->args(),
   );

   my ($out, $err);
   {
      local $SIG{TERM} = sub {
         WARN "cpanm: received TERM signal, ignoring";
      };
      INFO "calling @command";
      IPC::Run::run \@command, \undef, \*STDOUT, \*STDERR
        or LOGDIE "cpanm: $? ($err)";
   }

   INFO 'onboarding completed, indexing...';
   $self->_do_index($target);
   my $data_for = $self->last_index();

   INFO 'saving distlist';
   my @distros = $self->last_distlist();
   $self->_save2($target->file('distlist.txt'), join "\n", @distros, '');

   INFO 'saving modlist';
   my @modules = $self->last_modlist();
   $self->_save2($target->file('modlist.txt'), join "\n", @modules, '');

   my $file = $target->file('install.sh');
   if (!-e $file) {
      $self->_save2($file, <<'END_OF_INSTALL');
#!/bin/bash
ME=$(readlink -f "$0")
MYDIR=$(dirname "$ME")

TARGET="$MYDIR/local"
[ $# -gt 0 ] && TARGET=$1

if [ -n "$TARGET" ]; then
   "$MYDIR/cpanm" --mirror "file://$MYDIR" --mirror-only \
      -L "$TARGET" \
      $(<"$MYDIR/modlist.txt")
else
   "$MYDIR/cpanm" --mirror "file://$MYDIR" --mirror-only \
      $(<"$MYDIR/modlist.txt")
fi
END_OF_INSTALL
      chmod 0777 & ~umask(), $file->stringify();
   } ## end if (!-e $file)

   $file = $target->file('cpanm');
   if (!-e $file) {
      my $cpanm = which('cpanm');
      File::Copy::copy($cpanm, $file->stringify());
      chmod 0777 & ~umask(), $file->stringify();
   }
} ## end sub action_update

{
   no strict 'subs';
   *action_install = \&action_update;
   *action_add     = \&action_update;
}

sub action_inject {
   my ($self) = @_;

   my $target = $self->target_dir;
   $target->mkpath() unless -d $target;

   my $author = $self->config('author') // $ENV{EPAN_AUTHOR} // 'LOCAL';
   my $first = substr $author, 0, 1;
   my $first_two = substr $author, 0, 2;
   my $repo = $target->subdir(qw< authors id >, $first, $first_two, $author);
   $repo->mkpath;
   $repo = $repo->stringify;

   File::Copy::copy($_, $repo) for $self->args;

   INFO 'onboarding completed, indexing...';
   $self->_do_index($target);

   return;
}

sub _list_obsoletes {
   my ($self) = @_;
   my $basedir = $self->target_dir;
   my $data_for = $self->_collect_index_for($basedir);
   return sort {$a cmp $b} keys %{$data_for->{obsolete}};
}

sub action_list_obsoletes {
   my ($self) = @_;
   say for $self->_list_obsoletes;
   return;
}

sub action_purge_obsoletes {
   my ($self) = @_;
   for my $file ($self->_list_obsoletes) {
      INFO "removing $file";
      unlink $file;
   }
   return;
}

sub action_list_actions {
   my $self = shift;
   no strict 'refs';
   say 'Available actions:';
   say for
     sort {$a cmp $b}
     map {s/^action_/- /; s/_/-/g; $_ }
     grep {/^action_/ && $self->can($_)}
     keys %{ref($self)."::"};
   return;
}

sub last_distlist {
   my ($self) = @_;
   return keys %{$self->last_index()->{bare_distro}};
}

sub last_modlist {
   my ($self) = @_;
   my @retval =
     map { (sort keys %$_)[0] }
     values %{$self->last_index()->{bare_distro}};
} ## end sub last_modlist

1;
__END__
